File: python/rate_limiter/implementation/token_bucket.py
import time
from threading import Lock


class TokenBucket:
    def __init__(
        self,
        burst_capacity: int,
        refill_rate: float,
        refill_interval: float = 1.0,
    ) -> None:
        if burst_capacity <= 0:
            raise ValueError("burst_capacity must be positive")
        if refill_rate <= 0:
            raise ValueError("refill_rate must be positive")
        if refill_interval <= 0:
            raise ValueError("refill_interval must be positive")

        self._capacity = burst_capacity
        self._refill_rate = refill_rate
        self._refill_interval = refill_interval
        self._tokens: float = float(burst_capacity)
        self._last_refill = time.monotonic()
        self._lock = Lock()

    def allow_request(self, tokens: int = 1) -> bool:
        if tokens <= 0:
            raise ValueError("tokens must be positive")

        with self._lock:
            self._refill()
            if self._tokens >= tokens:
                self._tokens -= tokens
                return True
            return False

    def _refill(self) -> None:
        now = time.monotonic()
        elapsed = now - self._last_refill
        if elapsed >= self._refill_interval:
            intervals_passed = elapsed / self._refill_interval
            added = intervals_passed * self._refill_rate
            self._tokens = min(self._capacity, self._tokens + added)
            self._last_refill = now

    @property
    def available_tokens(self) -> float:
        with self._lock:
            self._refill()
            return self._tokens

    @property
    def capacity(self) -> int:
        return self._capacity

    def reset(self) -> None:
        with self._lock:
            self._tokens = self._capacity
            self._last_refill = time.monotonic()

File: java/rate_limiter/src/main/java/com/qe/test/ratelimit/TokenBucket.java
package com.qe.test.ratelimit;

import java.util.concurrent.locks.ReentrantLock;

public class TokenBucket {
    private final long burstCapacity;
    private final double refillRate;
    private double tokens;
    private long lastRefillTimestamp;
    private final ReentrantLock lock = new ReentrantLock();

    public TokenBucket(long burstCapacity, double refillRate) {
        if (burstCapacity <= 0) {
            throw new IllegalArgumentException("burstCapacity must be positive: " + burstCapacity);
        }
        if (refillRate <= 0) {
            throw new IllegalArgumentException("refillRate must be positive: " + refillRate);
        }
        this.burstCapacity = burstCapacity;
        this.refillRate = refillRate;
        this.tokens = burstCapacity;
        this.lastRefillTimestamp = System.nanoTime();
    }

    public boolean allowRequest() {
        lock.lock();
        try {
            refill();
            if (tokens >= 1.0) {
                tokens -= 1.0;
                return true;
            }
            return false;
        } finally {
            lock.unlock();
        }
    }

    public boolean allowRequest(long tokens) {
        if (tokens <= 0) {
            throw new IllegalArgumentException("tokens must be positive: " + tokens);
        }
        lock.lock();
        try {
            refill();
            if (this.tokens >= tokens) {
                this.tokens -= tokens;
                return true;
            }
            return false;
        } finally {
            lock.unlock();
        }
    }

    public double getAvailableTokens() {
        lock.lock();
        try {
            refill();
            return tokens;
        } finally {
            lock.unlock();
        }
    }

    private void refill() {
        long now = System.nanoTime();
        double elapsed = (now - lastRefillTimestamp) / 1_000_000_000.0;
        tokens = Math.min(burstCapacity, tokens + elapsed * refillRate);
        lastRefillTimestamp = now;
    }
}
